File: src/main/java/akostenko/aicars/race/Player.java
package akostenko.aicars.race;

import static java.lang.StrictMath.abs;
import static java.lang.StrictMath.min;

public class Player extends Driver {

    private final double steeringSensitivity = 4;
    private final double accelerationSensitivity = 5;
    private final double breakingSensitivity = 10;
    private final double fullInputTime = 0.5;

    @Override
    public String getName() {
        return "Player";
    }

    private double accelerating;
    private double breaking;
    private double steering;

    public void accelerate(boolean apply, double ms) {
        accelerating += ms/1000 * fullInputTime*accelerationSensitivity * (apply ? 1 : -3);
        accelerating = accelerating < 0 ? 0 : accelerating > 1 ? 1 : accelerating;
    }

    public void breaks(boolean apply, double ms) {
        breaking += ms/1000 * fullInputTime*breakingSensitivity * (apply ? 1 : -3);
        breaking = breaking < 0 ? 0 : breaking > 1 ? 1 : breaking;
    }

    public void turn(boolean left, boolean right, double ms) {
        double steeringDelta = ms / 1000 * fullInputTime * steeringSensitivity;
        double turnLeftDelta = steeringDelta * (left ? -1 : 0) * (steering > 0 ? 2 : 1);
        double turnRightDelta = steeringDelta * (right ? +1 : 0) * (steering < 0 ? 2 : 1);
        steering += turnLeftDelta + turnRightDelta;
        if (!left && !right) {
            // inertia of steering
            steeringDelta = min(steeringDelta, abs(steering));
            steeringDelta = steering > 0 ? -steeringDelta : (steering < 0 ? steeringDelta : 0);
            steering += steeringDelta;
        } else {
            steering = steering < -1 ? -1 : steering > 1 ? 1 : steering;
        }
    }

    @Override
    public double accelerating() {
        return accelerating;
    }

    @Override
    public double breaking() {
        return breaking;
    }

    @Override
    public double steering() {
        return steering;
    }

    @Override
    public void update(double seconds) {

    }
}

File: src/main/java/akostenko/aicars/race/Player.kt
package akostenko.aicars.race

import java.lang.StrictMath.abs
import java.lang.StrictMath.min

class Player : Driver() {

    private val steeringSensitivity = 4.0
    private val accelerationSensitivity = 5.0
    private val breakingSensitivity = 10.0
    private val fullInputTime = 0.5

    override val name: String
        get() = "Player"

    private var accelerating: Double = 0.toDouble()
    private var breaking: Double = 0.toDouble()
    private var steering: Double = 0.toDouble()

    fun accelerate(apply: Boolean, ms: Double) {
        accelerating += ms / 1000 * fullInputTime * accelerationSensitivity * (if (apply) 1 else -3).toDouble()
        accelerating = if (accelerating < 0) 0 else if (accelerating > 1) 1 else accelerating
    }

    fun breaks(apply: Boolean, ms: Double) {
        breaking += ms / 1000 * fullInputTime * breakingSensitivity * (if (apply) 1 else -3).toDouble()
        breaking = if (breaking < 0) 0 else if (breaking > 1) 1 else breaking
    }

    fun turn(left: Boolean, right: Boolean, ms: Double) {
        var steeringDelta = ms / 1000 * fullInputTime * steeringSensitivity
        val turnLeftDelta = steeringDelta * (if (left) -1 else 0).toDouble() * (if (steering > 0) 2 else 1).toDouble()
        val turnRightDelta = steeringDelta * (if (right) +1 else 0).toDouble() * (if (steering < 0) 2 else 1).toDouble()
        steering += turnLeftDelta + turnRightDelta
        if (!left && !right) {
            // inertia of steering
            steeringDelta = min(steeringDelta, abs(steering))
            steeringDelta = if (steering > 0) -steeringDelta else if (steering < 0) steeringDelta else 0
            steering += steeringDelta
        } else {
            steering = if (steering < -1) -1 else if (steering > 1) 1 else steering
        }
    }

    override fun accelerating(): Double {
        return accelerating
    }

    override fun breaking(): Double {
        return breaking
    }

    override fun steering(): Double {
        return steering
    }

    override fun update(seconds: Double) {

    }
}
